stratum/hal/lib/common/p4_service.cc

```diff
@@ -536,13 +536,32 @@ ::grpc::Status P4Service::StreamChannel(
         break;
       }
       case ::p4::v1::StreamMessageRequest::kPacket: {
-        // If this stream is not the master stream do not do anything.
-        if (!IsMasterController(node_id, connection_id)) break;
-        // If master, try to transmit the packet. No error reporting.
+        // If this stream is not the master stream generate a stream error.
+        if (!IsMasterController(node_id, connection_id)) {
+          ::p4::v1::StreamMessageResponse resp;
+          auto stream_error = resp.mutable_error();
+          stream_error->set_canonical_code(
+              ::grpc::StatusCode::PERMISSION_DENIED);
+          stream_error->set_message("Controller is not master.");
+          *stream_error->mutable_packet_out()->mutable_packet_out() =
+              req.packet();
+          stream->Write(resp);  // Best effort.
```

Comment:
I think this code could be a function, like: https://github.com/stratum/stratum/blob/9f5bd2b285badbef11e81eca6c31d4a3c4342843/stratum/hal/lib/common/config_monitoring_service.cc#L337-L348

Reply:
sounds good. But I am not sure if we can share this code between gNMI and P4Runtime (maybe use the template?)

Reply:
It's not that easy, because we have to personalize the error before sending. PacketOut errors include the packet, for example.